### PR TITLE
fix(security): validate GitHub user during integration installation

### PR DESCRIPTION
**Pull request by oioki**
We're adding one more step in the GitHub integration installation pipeline, namely GitHub OAuth2 authorize. This is transparent from the UX perspective as the data exchange happens without user interaction.

The pipeline will now fail in these cases:
- If there is a mismatch between currently authenticated GitHub user (derived from OAuth2 authorize step) and the user who installed the GitHub app (https://github.com/apps/sentry-io)
- If there is a mismatch between `state` parameter supplied by user and pipeline signature
- If GitHub could not generate correct `access_token` from the `code` (wrong or attempt of re-use of `code`).

In all those cases, this error is shown:
image

### CODE REVIEW

File: src/sentry/integrations/github/integration.py [hunk 1]
```diff
@@ -426,11 +423,76 @@ def dispatch(self, request: Request, pipeline: Pipeline) -> HttpResponse:
...
+        try:
+            body = safe_urlread(req).decode("utf-8")
+            payload = dict(parse_qsl(body))
+        except Exception:
+            payload = {}
+
+        if "access_token" not in payload:
+            return self.error(request)
+
+        # Check that the authenticated GitHub user is the same as who installed the app.
```
Comment (nhsiehgit) [inline src/sentry/integrations/github/integration.py:490]: 👌 nice

File: src/sentry/integrations/github/integration.py [hunk 2]
```diff
@@ -426,11 +423,76 @@ def dispatch(self, request: Request, pipeline: Pipeline) -> HttpResponse:
...
                 },
                 request=request,
             )
 
         # OrganizationIntegration does not exist, but Integration does exist.
         pipeline.bind_state("installation_id", request.GET["installation_id"])
         return pipeline.next_step()
+
+
+class GitHubUserValidation(PipelineView):
```
Comment (nhsiehgit) [inline src/sentry/integrations/github/integration.py:436]: Can we write some tests for this validator?

for each step in the dispatch where we throw an error, we could mock the state of the client and assert that the error is thrown?

**Review by nhsiehgit — COMMENTED** [review]
Nice this looks pretty good

can we validate this in a dev-env / record the flow?

File: src/sentry/integrations/github/integration.py [hunk 3]
```diff
@@ -118,6 +123,12 @@ def build_repository_query(metadata: Mapping[str, Any], name: str, query: str) -
     return f"{account_type}:{name} {query}".encode()
 
 
+def get_document_origin(org) -> str:
```
[the remark this answers is not in the document]
Comment (oioki) [reply src/sentry/integrations/github/integration.py:126]: Yes, just refactoring to de-duplicate the code.

**Comment by oioki** [comment]
Closing in favor of https://github.com/getsentry/sentry/pull/67876

**Pull request closed without merging**